### PR TITLE
Conditionally Borrowed Ranges

### LINKED ISSUE #1682: LWG-3494 Allow ranges to be conditionally borrowed
LWG-3494 Allow ranges to be conditionally borrowed
was resolved by:
[P2017R1](https://wg21.link/P2017R1) Conditionally Borrowed Ranges

Therefore, this is an LWG issue that should be applied to C++20, not a C++23 feature as I initially thought.

### PR DESCRIPTION
**Pull request by AdamBucior**
Fixes #1682.

### CODE REVIEW

**Review by miscco — APPROVED** [review]
Looks good to me!

Did you verify that we actually test against a borrowed range?

**Comment by AdamBucior** [comment]
> Did you verify that we actually test against a borrowed range?

I don't have a deep understanding of the ranges test stuff, but implementing this forced me to fix the common view test, so it seems we do.

**Comment by miscco** [comment]
> > Did you verify that we actually test against a borrowed range?
> 
> I don't have a deep understanding of the ranges test stuff, but implementing this forced me to fix the common view test, so it seems we do.

As long as we use a span, we should be fine

**Comment by miscco** [comment]
> > > Did you verify that we actually test against a borrowed range?
> > 
> > 
> > I don't have a deep understanding of the ranges test stuff, but implementing this forced me to fix the common view test, so it seems we do.
> 
> As long as we use a span, we should be fine

We need to add borrowed range support via a span to take_view and drop_view the other already test it

**Review by miscco — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Resolved trivial merge conflict (adjacent adds, kept both) in `tests/std/tests/P0896R4_views_take/test.cpp`.

**Comment by StephanTLavavej** [comment]
The tests are failing with:
```
C:\a\1\s\tests\std\tests\P0896R4_views_elements\test.cpp(50): error C2338: borrowed_range<R> == borrowed_range<Rng>
C:\a\1\s\tests\std\tests\P0896R4_views_elements\test.cpp(423): note: see reference to function template instantiation 'bool test_one<const std::ranges::transform_view<std::ranges::iota_view<_Ty1,_Ty2>,_Ty>&>(Rng)' being compiled
        with
        [
            _Ty1=unsigned __int64,
            _Ty2=size_t,
            _Ty=main::<lambda_1>,
            Rng=const std::ranges::transform_view<std::ranges::iota_view<unsigned __int64,size_t>,main::<lambda_1>> &
        ]
```

**Review by CaseyCarter — APPROVED** [review]

File: tests/std/tests/P0896R4_views_take/test.cpp [hunk 1]
```diff
@@ -82,7 +82,7 @@ concept CanViewTake = requires(Rng&& r) {
 template <ranges::input_range Rng, ranges::random_access_range Expected>
 constexpr bool test_one(Rng&& rng, Expected&& expected) {
     using ranges::input_range, ranges::forward_range, ranges::bidirectional_range, ranges::random_access_range,
-        ranges::contiguous_range;
+        ranges::contiguous_range, ranges::borrowed_range;
     using ranges::take_view, ranges::common_range, ranges::enable_borrowed_range, ranges::iterator_t, ranges::prev,
         ranges::range, ranges::sentinel_t, ranges::sized_range;
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_take/test.cpp:87]: Nitpick: There are two lists of using-declarations here. The first list is for various "strengths" of ranges, while the second list is for other machinery. To be consistent with this organization, `ranges::borrowed_range` should be added to the "other machinery" list.

(Compare `P0896R4_views_drop` which is following the existing organization, although its lists occur in the opposite order.)

**Review by StephanTLavavej — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for implementing this LWG issue resolution! :rocket: :heart_eyes_cat: :tada:

**Pull request merged**